### PR TITLE
 Improve single vs. multi-touch zoom & pan interaction (#7196)

### LINKED ISSUE #6900: Panning not possible after pinch zoom without lifting finger
**mapbox-gl-js version**:
v45
**browser**:
Chrome Android

### Steps to Trigger Behavior

 1. Pinch zoom in and out, but leave fingers on screen
 2. Lift one finger, try and pan with the finger still on the screen
 3. Map does not move

### Link to Demonstration
Will try and film this but believe the replication steps are easy to follow

### Expected Behavior
Google Maps and most other touch slippy maps seem to detect one finger has been lifted and allow pans seamlessly after a pinch zoom.

### Actual Behavior
Map remains fixed regardless of how the remaining finger is moved on the screen.

### LINKED ISSUE #7196: Single finger pan events overlaps with multi touch pan
There is no touch point count filtering in single finger pan. When you start multi touch pan there’s single and multi touch pan functions doing DOM updates simultaneously.

I noticed this when was fixing the multi touch center issue. Will debug more if there’s an easy fix. I’m not sure how the events should work in such case, so didn’t want to rush with a fix..

### PR DESCRIPTION
**Pull request by vakila**
Improves interaction when changing between single- and multi-finger touches, by checking for multi-touch events in `dragPan` handler and deactivating/reactivating that handler as appropriate if `touchZoomRotate` handler is active. 

## Changes
- Stop an active single-touch if the user adds additional fingers and starts rotating/zooming (i.e. if touchZoomRotate is enabled and it begins a rotate/zoom, dragPan stops the drag) (fixes #7196) If the user only pans but does not zoom/rotate with multiple fingers, dragPan continues the drag.
- Start a single-touch drag if the user removes all but one finger (fixes #6900)
- Add tests for expected single vs. multi-touch interactions
  - change simulated `touchend`/`touchcancel` events to have an empty `touches` list by default

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - ~document any changes to public APIs~ N/A
 - ~post benchmark scores~ N/A for these changes? 
 - [x] manually test the debug page - manually tested touch interactions on Android/Chrome

cc @asheemmamoowala

### CODE REVIEW

**Comment by vakila** [comment]
To test the behavior changes implemented here (cc @asheemmamoowala )

1. While single-finger panning, add a second finger(s) and continue panning/zooming
   - Before: the map still pans/zooms, but it's stuttery and not smooth (because both `dragPan` and `touchZoomRotate` handlers are simultaneously trying to move the map)
   - After (with PR): panning/zooming continues, more smoothly (because only `touchZoomRotate` is active, `dragPan` has been deactivated on the multi-finger `touchstart`)

2. Remove the second finger leaving one finger on the screen, and try to keep panning.
   - Before: panning/interaction stops entirely until you end the single-finger touch. (because the multi-finger `touchend` deactivated both `touchZoomRotate` and `dragPan`, even though there's still one finger on the map)
   - After: single-finger panning continues as expected (because `dragPan` has been re-activated on the multi-finger `touchend` as if it were a single-finger `touchstart`).

**Comment by vakila** [comment]
Ah just discovered the multi-to-single-touch interaction problem was ticketed in #6900, which this also fixes; updated description above.

**Comment by 1ec5** [comment]
Slightly off-topic: does GL JS currently address the case of a two-finger pinch-to-zoom gesture where the averaged point between the fingers moves (a “wandering zoom”)? On a touchscreen, it’s quite difficult to keep that centerpoint stationary as you pinch your fingers or spread them apart. To maintain a responsive feel, it’s necessary to pan at the same time as zooming (though perhaps some use cases may call for a more precise gesture handler that locks the center while zooming).

**Comment by mourner** [comment]
@1ec5 currently not, and it's been bothering me too — perhaps we should capture that in a ticket too. Leaflet does this.

File: src/ui/handler/drag_pan.js [hunk 1]
```diff
@@ -231,7 +242,29 @@ class DragPanHandler {
         }
     }
 
-    _onBlur(e: FocusEvent) {
+    _onMultiTouchEnd(e: TouchEvent) {
+        // A multi-finger touch is ending; decide whether to activate dragging
+
+        if (e.touches && e.touches.length > 1) return; // there are still multiple fingers touching; wait
```
Comment (arindam1993) [inline src/ui/handler/drag_pan.js]: Maybe move this logic to something like `isMultiTouch(e)`?

**Comment by vakila** [comment]
Thanks for the review @arindam1993! Did you by any chance try this out on a touch device, and if so did you have any issues with the new behavior as describe above (https://github.com/mapbox/mapbox-gl-js/pull/8100#issuecomment-486329653)? Wondering if you ran into the same issues @asheemmamoowala had a while back.

**Comment by vakila** [comment]
This is ready to go! I got the remaining tests passing and as far as @arindam1993 and I can tell through debug page testing on Android, iOS, & Surface Book, the behavior is as intended. @asheemmamoowala  and/or @arindam1993 can you please :eyes: so we can (finally!) get this merged in? :pray: :heart: :crossed_fingers:

File: src/ui/handler/drag_pan.js [hunk 2]
```diff
@@ -147,28 +154,39 @@ class DragPanHandler {
 
         this._state = 'pending';
         this._startPos = this._mouseDownPos = this._prevPos = this._lastPos = DOM.mousePos(this._el, e);
+        this._startTouch = this._lastTouch = e instanceof window.TouchEvent ? DOM.touchPos(this._el, e) : null;
```
Comment (arindam1993) [inline src/ui/handler/drag_pan.js]: import `window`?

File: src/ui/handler/drag_pan.js [hunk 3]
```diff
@@ -147,28 +154,39 @@ class DragPanHandler {
...
         this._inertia = [[browser.now(), this._startPos]];
     }
 
+    _touchesMatch(lastTouch: ?Array<Point>, thisTouch: ?Array<Point>) {
+        if (!lastTouch || !thisTouch || lastTouch.length !== thisTouch.length) return false;
+        for (let i = 0; i < lastTouch.length; i++) {
+            if (!lastTouch[i].equals(thisTouch[i])) return false;
+        }
+        return true;
```
Comment (arindam1993) [inline src/ui/handler/drag_pan.js]: small thing, but you could use `.every` to clean it up a bit.

File: src/ui/handler/drag_pan.js (hunk 2 above)
Comment (vakila) [reply src/ui/handler/drag_pan.js]: @arindam1993 maybe I'm misunderstanding what you mean but we import `window` in line 5

Comment (arindam1993) [reply src/ui/handler/drag_pan.js]: No you're not, my derp 🤦‍♂

File: src/ui/handler/drag_pan.js (hunk 3 above)
Comment (vakila) [reply src/ui/handler/drag_pan.js]: Fixed in https://github.com/mapbox/mapbox-gl-js/pull/8100/commits/b50f108fa85748c6c139a3e2fb4f237305b2a9fe

**Comment by vakila** [comment]
Thanks for the :heavy_check_mark: @arindam1993! Made one update per your comment. 

@asheemmamoowala did you want to :eyes: this again or are we good to :ship: ?

**Pull request merged**